### ISSUE TITLE
Q: defer() in more general setting

### ISSUE THREAD

**Issue by ib00** [mortie/snow#31]
I am marveling at your defer() implementation. Very clever trick!

I am wondering if defer() statement could be used as a general programming tool, ie. not in the context of Snow.
What would have to be changed?

**Comment by mortie** [comment]
The main problem with defer in general code is that you need to run code when a scope exits. In Snow, I know when that is; it's when a test fails or succeeds, but in normal C, you may return, break, continue, reach the end of a loop, reach the end of a function, ...

You also need to run some setup code at the beginning of a scope.

One solution would be to have a macro which has to be at the beginning of a scope which sets up any data structures you need and which uses the GNU extension `__attribute__((cleanup))` to run some code when the scope exits. Or, for an ISO C compliant solution, you can introduce your own flow control macros there's a few defer libraries which do this (for example [this one](https://github.com/moon-chilled/Defer)).

Another alternative is to have a macro which works like a statement (such as `deferred_scope`) which uses my for loop hack; this would make functions look like this:

```
void whatever() {
    deferred_scope {
        void *whatever = malloc(100);
        defer(free(whatever));
    }
}
```

Basically, there are many potential ways to make a generic defer implementation, but none are extremely ergonomic.

Another issue you'll encounter is that if you use a longjmp-based solution, there are some rather unintuitive limitations to what you can do between the `setjmp` and the `longjmp`. These limitations may not apply if you use GNU's label pointer extension, I'm not quite sure.

**Comment by ib00** [comment]
Thanks!
